[Woo POS][Product Search] Analytics

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: WOOPRD-465
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

[Specs](https://wp.me/pdfdoF-6Wm)

This PR adds tracking for the following events:

- Log `search_button_tapped` with `item_list_type parameter: products`
- Update `item_added_to_cart` to add `source` parameter: `list | search_result`
- Add `pre_search_recent_term_tapped` event, with `item_list_type parameter: products`
- Add `items_next_page_loaded` event with `item_list_type and search: bool parameters`

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
N/A

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->
Verify in logcat that the above events are tracked.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Verified that the events are tracked.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
```bash
🔵 Tracked: woocommerceandroid_pos_search_button_tapped, Properties: {"item_list_type":"products"}
🔵 Tracked: woocommerceandroid_pos_item_added_to_cart, Properties: {"source":"search_result"}
🔵 Tracked: woocommerceandroid_pos_pre_search_recent_term_tapped, Properties: {"item_list_type":"products"}
🔵 Tracked: woocommerceandroid_pos_item_added_to_cart, Properties: {"source":"pre_search_list"}
🔵 Tracked: woocommerceandroid_pos_item_added_to_cart, Properties: {"source":"search_result","product_type":"variation"}
```

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Add new event objects

samiuelson committed:
Track SearchButtonTapped

samiuelson committed:
Clean up code

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Add source param to ItemAddedToCart event

samiuelson committed:
Add pre_search_list source param to ItemAddedToCart event

samiuelson committed:
Make `source` property dynamic in variations VM

Track ItemClicked event in variations screen with real source (products, search)

samiuelson committed:
Update tests

samiuelson committed:
Fix coroutine cancellations in test

samiuelson committed:
Clean up tests

samiuelson committed:
Track ItemsNextPageLoaded

samiuelson committed:
Track PreSearchRecentTermTapped

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Merge branch 'trunk' into wooprd-465-woo-posproduct-search-implement-tracking

samiuelson committed:
Merge branch 'trunk' into wooprd-465-woo-posproduct-search-implement-tracking

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/WooPosItemsSearchHelper.kt:
```diff
@@ -165,6 +174,25 @@ class WooPosItemsSearchHelper @Inject constructor(
+            .map { it is SearchState.Visible && it.state is WooPosSearchInputState.Open }
+            .distinctUntilChanged()
+            .filter { it }
+            .onEach {
```
onEach is invoked only when WooPosSearchInputState changes from `Closed` to `Open`.

samiuelson committed:
Track recent search selected

samiuelson committed:
Add test

samiuelson committed:
Clean up code

samiuelson committed:
Clean up code

samiuelson committed:
Merge `trunk` and resolve conflicts

samiuelson committed:
Merge branch 'trunk' into wooprd-465-woo-posproduct-search-implement-tracking

samiuelson committed:
Improve test name

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeChildToParentCommunication.kt:
```diff
@@ -25,7 +26,10 @@ sealed class ChildToParentEvent {
-    data class ItemClickedInProductSelector(val itemData: WooPosItemsViewModel.ItemClickedData) : ChildToParentEvent()
+    data class ItemClickedInProductSelector(
+        val itemData: WooPosItemsViewModel.ItemClickedData,
+        val source: WooPosItemSource = WooPosItemSource.PRODUCT_LIST
```
📓 I think it's a bit risky to define the default value here - I think it's safer to force the developer to explicitly mention which place they are calling the event from.

samiuelson replied:
Good catch; removed: b5db1965be83b9408423b22fe80238fe63c7cbd0

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/items/WooPosItemsSearchHelper.kt:
```diff
@@ -165,6 +174,25 @@ class WooPosItemsSearchHelper @Inject constructor(
         )
     }
 
+    private fun observeAndTrackSearchInputStateOpen(
```
This code feels complex for what it is doing. Also, it might be easy to miss bugs for states like:
1. Open Search - event tracked
2. Add item to the cart
3. Tap on checkout (notice we are not closing search)
4. Complete order
5. Tap on New Order
6. Tap on Search - technically, the search might still be open from the previous order.

I mean this flow works at the moment, however it is due to a side effect (we are clearing order and search state), it slightly indicates the solution isn't very robust. Have you considered passing a `tap` event from the UI layer and logging explicit taps.

samiuelson replied:
Makes sense. I added explicit event: ea471d11285c7a827420b94f2d9128c51eba4da9 and test d9bfeaf83ec4297a3d3790189a33fc177af57000

malinajirka reviewed (approved):
Thanks @samiuelson!

The description doesn't mention `"source":"pre_search_list"` but I guess it is desired we track that, right?

Other than that, all the tracking seems to be working as expected. I've left a couple non-blocking comments.

samiuelson committed:
Remove default arg value

samiuelson committed:
Use const instead of hardcoded string

samiuelson committed:
Simplify search icon tap detection and tracking

samiuelson committed:
Update tests

samiuelson committed:
Merge branch 'trunk' into wooprd-465-woo-posproduct-search-implement-tracking

samiuelson committed:
Merge remote-tracking branch 'origin/trunk' into wooprd-465-woo-posproduct-search-implement-tracking

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/component/WooPosSearchInput.kt

samiuelson committed:
Clean up imports
